<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Plotting CellPhoneDB results • ktplots</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Plotting CellPhoneDB results">
<meta property="og:description" content="ktplots">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ktplots</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette.html">Plotting CellPhoneDB results</a>
    </li>
    <li>
      <a href="../articles/vignette_v5.html">New CellPhoneDB v5 results</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Plotting CellPhoneDB results</h1>
            
            <h4 data-toc-skip class="date">2023-08-18</h4>
      
      
      <div class="hidden name"><code>vignette.rmd</code></div>

    </div>

    
    
<p><a href="https://opensource.org/licenses/MIT" class="external-link"><img src="https://img.shields.io/badge/License-MIT-yellow.svg" alt="License: MIT"></a> <a href="https://codecov.io/gh/zktuong/ktplots" class="external-link"><img src="https://codecov.io/gh/zktuong/ktplots/branch/master/graph/badge.svg" alt="codecov"></a> <a href="https://github.com/zktuong/ktplots/actions/workflows/r.yml" class="external-link"><img src="https://github.com/zktuong/ktplots/actions/workflows/r.yml/badge.svg" alt="R"></a> <a href="https://doi.org/10.5281/zenodo.5717922" class="external-link"><img src="https://zenodo.org/badge/DOI/10.5281/zenodo.6728350.svg" alt="DOI"></a></p>
<div class="section level2">
<h2 id="ktplots">ktplots<a class="anchor" aria-label="anchor" href="#ktplots"></a>
</h2>
<p>Welcome to <code>ktplots</code>! This is a R package to help
visualise <code>CellPhoneDB</code> results. Here, we will go through a
quick tutorial on how to use the functions in this package.</p>
<p>For a python port of <code>ktplots</code>, please check out my other
<a href="https://www.github.com/zktuong/ktplotspy" class="external-link">repository</a>.</p>
<div class="section level3">
<h3 id="installation-instructions">Installation instructions<a class="anchor" aria-label="anchor" href="#installation-instructions"></a>
</h3>
<p>You can install the package via
<code><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">devtools::install_github()</a></code> function in R</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"devtools"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"zktuong/ktplots"</span>, dependencies <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="usage-instructions">Usage instructions<a class="anchor" aria-label="anchor" href="#usage-instructions"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ktplots</span><span class="op">)</span></span></code></pre></div>
<p>There is a test dataset in <code>SingleCellExperiment</code> format
to test the functions.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">kidneyimmune</span><span class="op">)</span></span>
<span><span class="co"># Some functions accept Seurat objects too.</span></span></code></pre></div>
<p>The data is downsampled from the <a href="https://kidneycellatlas.org" class="external-link">kidney cell atlas</a>.</p>
<p>For more info, please see <a href="https://science.sciencemag.org/content/365/6460/1461" class="external-link">Stewart et
al. kidney single cell data set published in Science 2019</a>.</p>
</div>
<div class="section level3">
<h3 id="prepare-input">Prepare input<a class="anchor" aria-label="anchor" href="#prepare-input"></a>
</h3>
<p>We will need 3 files to use this package, a
<code>SingleCellExperiment</code> (or <code>Seurat</code>; some
functions only accepts the former) object that correspond to the object
you used for <code>CellPhoneDB</code> and the <code>means.txt</code> and
<code>pvalues.txt</code> output. If you are using results from
<code>CellPhoneDB</code> <code>deg_analysis</code> mode from version
&gt;= 3, the <code>pvalues.txt</code> is
<code>relevant_interactions.txt</code> and also add
<code>degs_analysis = TRUE</code> into all the functions below.
<code>deconvoluted</code> is only used for
<code>plot_cpdb2/3/4</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># pvals &lt;- read.delim("pvalues.txt", check.names = FALSE)</span></span>
<span><span class="co"># means &lt;- read.delim("means.txt", check.names = FALSE)</span></span>
<span><span class="co"># decon = pd.read_csv("deconvoluted.txt", sep="\t")</span></span>
<span></span>
<span><span class="co"># I've provided an example datasets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">cpdb_output</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">cpdb_output2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="heatmap">Heatmap<a class="anchor" aria-label="anchor" href="#heatmap"></a>
</h3>
<p>The original heatmap plot from <code>CellPhoneDB</code> can be
achieved with this reimplemented function.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_cpdb_heatmap.html">plot_cpdb_heatmap</a></span><span class="op">(</span>pvals <span class="op">=</span> <span class="va">pvals2</span>, cellheight <span class="op">=</span> <span class="fl">10</span>, cellwidth <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>The current heatmap is directional (check <code>count_network</code>
and <code>interaction_edges</code> for more details in
<code>return_tables = True</code>).</p>
<p>To obtain the heatmap where the interaction counts are not
symmetrical, do:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_cpdb_heatmap.html">plot_cpdb_heatmap</a></span><span class="op">(</span>pvals <span class="op">=</span> <span class="va">pvals2</span>, cellheight <span class="op">=</span> <span class="fl">10</span>, cellwidth <span class="op">=</span> <span class="fl">10</span>, symmetrical <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>The values for the <code>symmetrical=FALSE</code> mode follow the
direction of the L-R direction where it’s always moleculeA:celltypeA
-&gt; moleculeB:celltypeB.</p>
<p>Therefore, if you trace on the <code>x-axis</code> for
<code>celltype A</code> [MNPa(mono)] to <code>celltype B</code> [CD8T
cell] on the <code>y-axis</code>:</p>
<p>A -&gt; B is 18 interactions</p>
<p>Whereas if you trace on the <code>y-axis</code> for
<code>celltype A</code> [MNPa(mono)] to <code>celltype B</code> [CD8T
cell] on the <code>x-axis</code>:</p>
<p>A -&gt; B is 9 interactions</p>
<p><code>symmetrical=TRUE</code> mode will return 18+9 = 27</p>
</div>
<div class="section level3">
<h3 id="dot-plot">Dot plot<a class="anchor" aria-label="anchor" href="#dot-plot"></a>
</h3>
<div class="section level4">
<h4 id="plot_cpdb">plot_cpdb<a class="anchor" aria-label="anchor" href="#plot_cpdb"></a>
</h4>
<p>A simple usage of <code>plot_cpdb</code> is as follows:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_cpdb.html">plot_cpdb</a></span><span class="op">(</span></span>
<span>    scdata<span class="op">=</span><span class="va">kidneyimmune</span>,</span>
<span>    cell_type1<span class="op">=</span><span class="st">"B cell"</span>,</span>
<span>    cell_type2<span class="op">=</span><span class="st">"."</span>,  <span class="co"># this means all cell-types</span></span>
<span>    celltype_key<span class="op">=</span><span class="st">"celltype"</span>,</span>
<span>    means<span class="op">=</span><span class="va">means2</span>,</span>
<span>    pvals<span class="op">=</span><span class="va">pvals2</span>,</span>
<span>    genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PTPRC"</span>, <span class="st">"TNFSF13"</span><span class="op">)</span>,</span>
<span>    title<span class="op">=</span><span class="st">"interacting interactions!"</span>,</span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-8-1.png" width="960"></p>
<p>Or don’t specify either and it will try to plot all significant
interactions.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_cpdb.html">plot_cpdb</a></span><span class="op">(</span></span>
<span>    scdata<span class="op">=</span><span class="va">kidneyimmune</span>,</span>
<span>    cell_type1<span class="op">=</span><span class="st">"B cell"</span>,</span>
<span>    cell_type2<span class="op">=</span><span class="st">"."</span>,</span>
<span>    celltype_key<span class="op">=</span><span class="st">"celltype"</span>,</span>
<span>    means<span class="op">=</span><span class="va">means2</span>,</span>
<span>    pvals<span class="op">=</span><span class="va">pvals2</span>,</span>
<span>    genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PTPRC"</span>, <span class="st">"CD40"</span>, <span class="st">"CLEC2D"</span><span class="op">)</span>,</span>
<span>    default_style<span class="op">=</span><span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-9-1.png" width="960"></p>
<p>you can also toggle options to <code>splitby_key</code> and
<code>gene_family</code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_cpdb.html">plot_cpdb</a></span><span class="op">(</span></span>
<span>    scdata <span class="op">=</span> <span class="va">kidneyimmune</span>,</span>
<span>    cell_type1 <span class="op">=</span> <span class="st">"B cell"</span>,</span>
<span>    cell_type2 <span class="op">=</span> <span class="st">"Neutrophil|MNPc|NK cell"</span>,</span>
<span>    celltype_key <span class="op">=</span> <span class="st">"celltype"</span>,</span>
<span>    means <span class="op">=</span> <span class="va">means</span>,</span>
<span>    pvals <span class="op">=</span> <span class="va">pvals</span>,</span>
<span>    splitby_key <span class="op">=</span> <span class="st">"Experiment"</span>,</span>
<span>    gene_family <span class="op">=</span> <span class="st">"chemokines"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-10-1.png" width="960"></p>
<p>if <code>genes</code> and <code>gene_family</code> are both not
specified, the function will try to plot everything.</p>
<p>Specifying <code>keep_significant_only</code> will only keep those
that are p&lt;0.05.</p>
<p>You can also specify more than 1 gene families:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_cpdb.html">plot_cpdb</a></span><span class="op">(</span></span>
<span>    scdata <span class="op">=</span> <span class="va">kidneyimmune</span>,</span>
<span>    cell_type1 <span class="op">=</span> <span class="st">"B cell"</span>,</span>
<span>    cell_type2 <span class="op">=</span> <span class="st">"CD4T cell"</span>,</span>
<span>    celltype_key <span class="op">=</span> <span class="st">"celltype"</span>,</span>
<span>    means <span class="op">=</span> <span class="va">means</span>,</span>
<span>    pvals <span class="op">=</span> <span class="va">pvals</span>,</span>
<span>    splitby_key <span class="op">=</span> <span class="st">"Experiment"</span>,</span>
<span>    gene_family <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Coinhibitory"</span>, <span class="st">"Costimulatory"</span><span class="op">)</span>,</span>
<span>    cluster_rows <span class="op">=</span> <span class="cn">FALSE</span> <span class="co"># ensures that the families are separate,</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-11-1.png" width="960"></p>
<p>And also provide custom families as a <code>data.frame</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>set1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CCR6"</span>, <span class="st">"CCL20"</span>, <span class="st">"CXCL10"</span>, <span class="st">"CCR3"</span>, <span class="st">"TNFRSF13C"</span><span class="op">)</span>, set2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CCL5"</span>, <span class="st">"CCR4"</span>, <span class="st">"PTPRC"</span>, <span class="st">"CD40"</span>, <span class="st">"CLEC2D"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_cpdb.html">plot_cpdb</a></span><span class="op">(</span></span>
<span>    scdata<span class="op">=</span><span class="va">kidneyimmune</span>,</span>
<span>    cell_type1 <span class="op">=</span> <span class="st">"B cell"</span>,</span>
<span>    cell_type2 <span class="op">=</span> <span class="st">"CD4T cell"</span>,</span>
<span>    celltype_key <span class="op">=</span> <span class="st">"celltype"</span>,</span>
<span>    means <span class="op">=</span> <span class="va">means</span>,</span>
<span>    pvals <span class="op">=</span> <span class="va">pvals</span>,</span>
<span>    splitby_key <span class="op">=</span> <span class="st">"Experiment"</span>,</span>
<span>    gene_family <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"set1"</span>, <span class="st">"set2"</span><span class="op">)</span>,</span>
<span>    custom_gene_family <span class="op">=</span> <span class="va">df</span>,</span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-12-1.png" width="960"></p>
</div>
<div class="section level4">
<h4 id="combine_cpdb">combine_cpdb<a class="anchor" aria-label="anchor" href="#combine_cpdb"></a>
</h4>
<p>For the <code>splitby_key</code> option to work, the annotation in
the meta file must be defined in the following format:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>{splitby_key}_{celltype_key}</span></code></pre></div>
<p>so to set up an example vector, it would be something like:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">kidneyimmune</span><span class="op">$</span><span class="va">Experiment</span>, <span class="st">"_"</span>, <span class="va">kidneyimmune</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span></code></pre></div>
<p>The recommended way to use <code>splitby_key</code> is to prepare the
data with <code>combine_cpdb</code> like in this example:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Assume you have 2 cellphonedb runs, one where it's just naive and the other is treated, you will end up with 2 cellphonedb out folders</span></span>
<span><span class="co"># remember, the celltype labels you provide to cellphonedb's meta.txt should already be like {splitby_key}_{celltype_key}</span></span>
<span><span class="co"># so the two meta.txt should look like:</span></span>
<span></span>
<span><span class="co"># naive file</span></span>
<span><span class="co"># ATTAGTCGATCGTAGT-1    naive_CD4Tcell</span></span>
<span><span class="co"># ATTAGTGGATCGTAGT-1    naive_CD4Tcell</span></span>
<span><span class="co"># ATTAGTCGACCGTAGT-1    naive_CD8Tcell</span></span>
<span><span class="co"># ATTAGTCGATCGTAGT-1    naive_CD8Tcell</span></span>
<span><span class="co"># ATGAGTCGATCGTAGT-1    naive_Bcell</span></span>
<span><span class="co"># ATTAGTCGATCGTGGT-1    naive_Bcell</span></span>
<span></span>
<span><span class="co"># treated file</span></span>
<span><span class="co"># ATTAGTCAATCGTAGT-1    treated_CD4Tcell</span></span>
<span><span class="co"># ATTAGTGGATCGTAGT-1    treated_CD4Tcell</span></span>
<span><span class="co"># ATTAGTCGACCATAGT-1    treated_CD8Tcell</span></span>
<span><span class="co"># ATTAGTAGATCGTAGT-1    treated_CD8Tcell</span></span>
<span><span class="co"># ATGAGTCGATCGTAAT-1    treated_Bcell</span></span>
<span><span class="co"># ATTAGTCGATCGTGAT-1    treated_Bcell</span></span>
<span></span>
<span><span class="co"># one you have set that up correctly, you can then read in the files.</span></span>
<span><span class="va">naive_means</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="st">"naive_out/means.txt"</span>, check.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">naive_pvals</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="st">"naive_out/pvalues.txt"</span>, check.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">naive_decon</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="st">"naive_out/deconvoluted.txt"</span>, check.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">treated_means</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="st">"treated_out/means.txt"</span>, check.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">treated_pvals</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="st">"treated_out/pvalues.txt"</span>, check.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">treated_decon</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="st">"treated_out/deconvoluted.txt"</span>, check.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">means</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combine_cpdb.html">combine_cpdb</a></span><span class="op">(</span><span class="va">naive_means</span>, <span class="va">treated_means</span><span class="op">)</span></span>
<span><span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combine_cpdb.html">combine_cpdb</a></span><span class="op">(</span><span class="va">naive_pvals</span>, <span class="va">treated_pvals</span><span class="op">)</span></span>
<span><span class="va">decon</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combine_cpdb.html">combine_cpdb</a></span><span class="op">(</span><span class="va">naive_decon</span>, <span class="va">treated_decon</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_cpdb.html">plot_cpdb</a></span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="plot_cpdb2">plot_cpdb2<a class="anchor" aria-label="anchor" href="#plot_cpdb2"></a>
</h4>
<p>Generates a circos-style wire/arc/chord plot for cellphonedb
results.</p>
<p>This function piggy-backs on the original <code>plot_cpdb</code>
function and generates the results like this:</p>
<p>Please help contribute to the interaction grouping list <a href="https://docs.google.com/spreadsheets/d/1O9OKU7J0NdeQNJAIMpsHtWAFvY014GDQ7aigdGUSTmc/edit?usp=sharing" class="external-link">here</a>!</p>
<p>Credits to Ben Stewart for coming up with the base code!</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/plot_cpdb2.html">plot_cpdb2</a></span><span class="op">(</span></span>
<span>    scdata <span class="op">=</span> <span class="va">kidneyimmune</span>,</span>
<span>    cell_type1 <span class="op">=</span> <span class="st">"B cell"</span>,</span>
<span>    cell_type2 <span class="op">=</span> <span class="st">"."</span>,</span>
<span>    celltype_key <span class="op">=</span> <span class="st">"celltype"</span>, <span class="co"># column name where the cell ids are located in the metadata</span></span>
<span>    means <span class="op">=</span> <span class="va">means2</span>,</span>
<span>    pvals <span class="op">=</span> <span class="va">pvals2</span>,</span>
<span>    deconvoluted <span class="op">=</span> <span class="va">decon2</span>, <span class="co"># new options from here on specific to plot_cpdb2</span></span>
<span>    desiredInteractions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4T cell"</span>, <span class="st">"B cell"</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B cell"</span>, <span class="st">"CD4T cell"</span><span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span>    interaction_grouping <span class="op">=</span> <span class="va">interaction_annotation</span>,</span>
<span>    edge_group_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="st">"Activating"</span> <span class="op">=</span> <span class="st">"#e15759"</span>,</span>
<span>        <span class="st">"Chemotaxis"</span> <span class="op">=</span> <span class="st">"#59a14f"</span>,</span>
<span>        <span class="st">"Inhibitory"</span> <span class="op">=</span> <span class="st">"#4e79a7"</span>,</span>
<span>        <span class="st">"Intracellular trafficking"</span> <span class="op">=</span> <span class="st">"#9c755f"</span>,</span>
<span>        <span class="st">"DC_development"</span> <span class="op">=</span> <span class="st">"#B07aa1"</span>,</span>
<span>        <span class="st">"Unknown"</span> <span class="op">=</span> <span class="st">"#e7e7e7"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    node_group_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="st">"CD4T cell"</span> <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>        <span class="st">"B cell"</span> <span class="op">=</span> <span class="st">"blue"</span></span>
<span>    <span class="op">)</span>,</span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-16-1.png" width="960"></p>
</div>
<div class="section level4">
<h4 id="plot_cpdb3">plot_cpdb3<a class="anchor" aria-label="anchor" href="#plot_cpdb3"></a>
</h4>
<p>Generates a chord diagram inspired from <a href="https://github.com/sqjin/CellChat" class="external-link">CellChat</a>’s way of showing
the data!</p>
<p>Usage is similar to <code>plot_cpdb2</code> but with reduced options.
Additional kwargs are passed to <code>plot_cpdb</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_cpdb3.html">plot_cpdb3</a></span><span class="op">(</span></span>
<span>    scdata <span class="op">=</span> <span class="va">kidneyimmune</span>,</span>
<span>    cell_type1 <span class="op">=</span> <span class="st">"B cell"</span>,</span>
<span>    cell_type2 <span class="op">=</span> <span class="st">"CD4T cell|MNPd"</span>,</span>
<span>    celltype_key <span class="op">=</span> <span class="st">"celltype"</span>, <span class="co"># column name where the cell ids are located in the metadata</span></span>
<span>    means <span class="op">=</span> <span class="va">means2</span>,</span>
<span>    pvals <span class="op">=</span> <span class="va">pvals2</span>,</span>
<span>    deconvoluted <span class="op">=</span> <span class="va">decon2</span> <span class="co"># new options from here on specific to plot_cpdb3</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-17-1.png" width="960"></p>
</div>
<div class="section level4">
<h4 id="plot_cpdb4">plot_cpdb4<a class="anchor" aria-label="anchor" href="#plot_cpdb4"></a>
</h4>
<p>New! Alternate way of showing the chord diagram for specific
interactions!</p>
<p>Usage is similar to <code>plot_cpdb3</code> but with additional
required <code>interaction</code> option. Additional kwargs are passed
to <code>plot_cpdb</code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_cpdb4.html">plot_cpdb4</a></span><span class="op">(</span></span>
<span>    scdata <span class="op">=</span> <span class="va">kidneyimmune</span>,</span>
<span>    interaction <span class="op">=</span> <span class="st">"CLEC2D-KLRB1"</span>,</span>
<span>    cell_type1 <span class="op">=</span> <span class="st">"NK"</span>,</span>
<span>    cell_type2 <span class="op">=</span> <span class="st">"Mast"</span>,</span>
<span>    celltype_key <span class="op">=</span> <span class="st">"celltype"</span>,</span>
<span>    means <span class="op">=</span> <span class="va">means2</span>,</span>
<span>    pvals <span class="op">=</span> <span class="va">pvals2</span>,</span>
<span>    deconvoluted <span class="op">=</span> <span class="va">decon2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-18-1.png" width="960"></p>
<p>or specify more than 1 interactions + only show specific cell-type
type interactions!</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_cpdb4.html">plot_cpdb4</a></span><span class="op">(</span></span>
<span>    interaction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CLEC2D-KLRB1"</span>, <span class="st">"CD40-CD40LG"</span><span class="op">)</span>,</span>
<span>    cell_type1 <span class="op">=</span> <span class="st">"NK|B"</span>, cell_type2 <span class="op">=</span> <span class="st">"Mast|CD4T"</span>,</span>
<span>    scdata <span class="op">=</span> <span class="va">kidneyimmune</span>,</span>
<span>    celltype_key <span class="op">=</span> <span class="st">"celltype"</span>,</span>
<span>    means <span class="op">=</span> <span class="va">means2</span>,</span>
<span>    pvals <span class="op">=</span> <span class="va">pvals2</span>,</span>
<span>    deconvoluted <span class="op">=</span> <span class="va">decon2</span>,</span>
<span>    desiredInteractions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NK cell"</span>, <span class="st">"Mast cell"</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NK cell"</span>, <span class="st">"NKT cell"</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NKT cell"</span>, <span class="st">"Mast cell"</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B cell"</span>, <span class="st">"CD4T cell"</span><span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span>    keep_significant_only <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="other-useful-functions">Other useful functions<a class="anchor" aria-label="anchor" href="#other-useful-functions"></a>
</h3>
<div class="section level4">
<h4 id="genedotplot">geneDotPlot<a class="anchor" aria-label="anchor" href="#genedotplot"></a>
</h4>
<p>Plotting gene expression dot plots heatmaps.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="co"># Note, this conflicts with tidyr devel version</span></span>
<span><span class="fu"><a href="../reference/geneDotPlot.html">geneDotPlot</a></span><span class="op">(</span></span>
<span>    scdata <span class="op">=</span> <span class="va">kidneyimmune</span>, <span class="co"># object</span></span>
<span>    genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD68"</span>, <span class="st">"CD80"</span>, <span class="st">"CD86"</span>, <span class="st">"CD74"</span>, <span class="st">"CD2"</span>, <span class="st">"CD5"</span><span class="op">)</span>, <span class="co"># genes to plot</span></span>
<span>    celltype_key <span class="op">=</span> <span class="st">"celltype"</span>, <span class="co"># column name in meta data that holds the cell-cluster ID/assignment</span></span>
<span>    splitby_key <span class="op">=</span> <span class="st">"Project"</span>, <span class="co"># column name in the meta data that you want to split the plotting by. If not provided, it will just plot according to celltype_key</span></span>
<span>    standard_scale <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>strip.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">0</span>, hjust <span class="op">=</span> <span class="fl">0</span>, size <span class="op">=</span> <span class="fl">7</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/misc.html">small_guide</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/misc.html">small_legend</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; data provided is a SingleCellExperiment/SummarizedExperiment object</span></span>
<span><span class="co">#&gt; extracting expression matrix</span></span>
<span><span class="co">#&gt; attempting to subset the expression matrix to the 6 genes provided</span></span>
<span><span class="co">#&gt; found 6 genes in the expression matrix</span></span>
<span><span class="co">#&gt; preparing the final dataframe ...</span></span>
<span><span class="co">#&gt; setting minimum percentage of cells expressing gene to be 5% of cluster/cell-type</span></span>
<span><span class="co">#&gt; the following genes are removed</span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h3>
<p>If you find these functions useful, please consider leaving a star,
citing this repository, and/or citing the following <a href="https://doi.org/10.5281/zenodo.5717922" class="external-link">DOI</a>:</p>
<p>To cite a specific version of <code>ktplots</code>, please follow the
links on the zenodo repository. e.g. v1.2.3:</p>
<pre><code>Zewen Kelvin Tuong. (2021). zktuong/ktplots: 1.2.3 (v1.2.3). Zenodo. https://doi.org/10.5281/zenodo.5717922</code></pre>
<p>Thank you!</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kelvin Tuong.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
